### DISCUSSION TITLE
FastAPI keeps booting new workers when deploying

### DISCUSSION THREAD

**Discussion by SansDK** [fastapi/fastapi#8310] [Questions]
Hi there,

I'm trying to run two different instances of FastAPI in the same Kubernetes pod (different Docker containers), but one of the two seems to keep booting up new workers. Therefor, the container uses a lot of CPU for nothing. It's also reporting that it's waiting for application startup constantly, but it is reachable (but slow).

Here is the log that I keep getting:

```
Checking for script in /app/prestart.sh
Running script /app/prestart.sh
Running inside /app/prestart.sh, you could add migrations to this file, e.g.:
 #! /usr/bin/env bash
 # Let the DB start
sleep 10;
# Run migrations
alembic upgrade head
 [2019-05-24 11:51:24 +0000] [1] [INFO] Starting gunicorn 19.9.0
[2019-05-24 11:51:24 +0000] [1] [INFO] Listening at: http://0.0.0.0:8081 (1)
[2019-05-24 11:51:24 +0000] [1] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2019-05-24 11:51:24 +0000] [8] [INFO] Booting worker with pid: 8
[2019-05-24 11:51:24 +0000] [9] [INFO] Booting worker with pid: 9
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
[2019-05-24 11:51:33 +0000] [9] [INFO] Started server process [9]
[2019-05-24 11:51:33 +0000] [9] [INFO] Waiting for application startup.
[2019-05-24 11:51:33 +0000] [8] [INFO] Started server process [8]
[2019-05-24 11:51:33 +0000] [8] [INFO] Waiting for application startup.
[2019-05-24 11:51:37 +0000] [18] [INFO] Booting worker with pid: 18
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
[2019-05-24 11:51:44 +0000] [18] [INFO] Started server process [18]
[2019-05-24 11:51:44 +0000] [18] [INFO] Waiting for application startup.
[2019-05-24 11:51:45 +0000] [23] [INFO] Booting worker with pid: 23
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
[2019-05-24 11:51:53 +0000] [23] [INFO] Started server process [23]
[2019-05-24 11:51:53 +0000] [23] [INFO] Waiting for application startup.
[2019-05-24 11:51:54 +0000] [28] [INFO] Booting worker with pid: 28
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
[2019-05-24 11:52:02 +0000] [28] [INFO] Started server process [28]
[2019-05-24 11:52:02 +0000] [28] [INFO] Waiting for application startup.
```

This will continue untill I shut it down. This problem weirdly does not happen when I run everything locally.

A difference between local and remote deployment is probably only the port that I set.
As you can see, I use port 8081:
```
[INFO] Listening at: http://0.0.0.0:8081 (1)
```

I set the port by specifying it in my Kubernetes yaml file, but I am unsure if this has something to do with the problem.

**Comment by euri10** [comment]
Iirc there are no max workers set so I'd either tweak the gunicorn conf to
set your number according to the server you deploy on, or set another env
var, can't remember on top of my head but it's in the docs I'm sure

Le ven. 24 mai 2019 à 2:03 PM, SansDK <notifications@github.com> a écrit :

> Hi there,
>
> I'm trying to run two different instances of FastAPI in the same
> Kubernetes pod (different Docker containers), but one of the two seems to
> keep booting up new workers. Therefor, the container uses a lot of CPU for
> nothing. It's also reporting that it's waiting for application startup
> constantly, but it is reachable (but slow).
>
> Here is the log that I keep getting:
>
> Checking for script in /app/prestart.sh
> Running script /app/prestart.sh
> Running inside /app/prestart.sh, you could add migrations to this file, e.g.:
>  #! /usr/bin/env bash
>  # Let the DB start
> sleep 10;
> # Run migrations
> alembic upgrade head
>  [2019-05-24 11:51:24 +0000] [1] [INFO] Starting gunicorn 19.9.0
> [2019-05-24 11:51:24 +0000] [1] [INFO] Listening at: http://0.0.0.0:8081 (1)
> [2019-05-24 11:51:24 +0000] [1] [INFO] Using worker: uvicorn.workers.UvicornWorker
> [2019-05-24 11:51:24 +0000] [8] [INFO] Booting worker with pid: 8
> [2019-05-24 11:51:24 +0000] [9] [INFO] Booting worker with pid: 9
> WARNING:root:email-validator not installed, email fields will be treated as str.
> To install, run: pip install email-validator
> WARNING:root:email-validator not installed, email fields will be treated as str.
> To install, run: pip install email-validator
> [2019-05-24 11:51:33 +0000] [9] [INFO] Started server process [9]
> [2019-05-24 11:51:33 +0000] [9] [INFO] Waiting for application startup.
> [2019-05-24 11:51:33 +0000] [8] [INFO] Started server process [8]
> [2019-05-24 11:51:33 +0000] [8] [INFO] Waiting for application startup.
> [2019-05-24 11:51:37 +0000] [18] [INFO] Booting worker with pid: 18
> WARNING:root:email-validator not installed, email fields will be treated as str.
> To install, run: pip install email-validator
> [2019-05-24 11:51:44 +0000] [18] [INFO] Started server process [18]
> [2019-05-24 11:51:44 +0000] [18] [INFO] Waiting for application startup.
> [2019-05-24 11:51:45 +0000] [23] [INFO] Booting worker with pid: 23
> WARNING:root:email-validator not installed, email fields will be treated as str.
> To install, run: pip install email-validator
> [2019-05-24 11:51:53 +0000] [23] [INFO] Started server process [23]
> [2019-05-24 11:51:53 +0000] [23] [INFO] Waiting for application startup.
> [2019-05-24 11:51:54 +0000] [28] [INFO] Booting worker with pid: 28
> WARNING:root:email-validator not installed, email fields will be treated as str.
> To install, run: pip install email-validator
> [2019-05-24 11:52:02 +0000] [28] [INFO] Started server process [28]
> [2019-05-24 11:52:02 +0000] [28] [INFO] Waiting for application startup.
>
> This will continue untill I shut it down. This problem weirdly does not
> happen when I run everything locally.
>
> A difference between local and remote deployment is probably only the port
> that I set.
> As you can see, I use port 8081:
>
> [INFO] Listening at: http://0.0.0.0:8081 (1)
>
> I set the port by specifying it in my Kubernetes yaml file, but I am
> unsure if this has something to do with the problem.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tiangolo/fastapi/issues/253?email_source=notifications&email_token=AAINSPWLEJ53IUGIRMY2MDDPW7KQBA5CNFSM4HPO4Y32YY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4GVV7SAQ>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAINSPTKBVSZMRPIDNLUCDTPW7KQBANCNFSM4HPO4Y3Q>
> .
>

**Comment by SansDK** [comment]
I'll have a look in the docs, but this isn't expected behavior.
I've deployed FastAPI with uvicorn and gunicorn before, and I never had to worry about gunicorn settings. Now it's just creating 1000+ workers for no reason and without any load at all.

For now I'll just try to set the ```WORKERS_PER_CORE``` environment variable and see if that fixes the problem. I guess it will, but it's still weird behavior because the default ```WORKERS_PER_CORE``` is 1, and I'm fairly sure I don't have a 1000+ cpu cores on my server.

Anyways, I'll report back if setting the variable fixes the problem. Thanks!

**Comment by SansDK** [comment]
Okay, I set the ```WORKERS_PER_CORE``` to 1 to no avail.
I struggle to understand how this is happening..
I have 1 Kubernetes pod with 2 different Docker container, each containing a FastAPI instance.
One of the two always works and the other one keeps doing this looping.
Since it keeps ```Waiting for application startup```, does it check on the wrong port for startup maybe?

What does the ```(1)``` mean on the line ```Listening at: http://0.0.0.0:8081 (1)```?

Thanks @euri10 for the help.

**Comment by tiangolo** [comment]
`WORKERS_PER_CORE` is a float telling the container how many workers per core to boot.

If you have it set to `1` and you have 64 processors with 2 cores each one (128 cores), it will start 128 workers.

To make it a fixed number of workers you can use `WEB_CONCURRENCY`: https://github.com/tiangolo/uvicorn-gunicorn-fastapi-docker#web_concurrency

That will fix the number of workers to that specific number (independent of the number of available cores).

The image is configured to, by default, get the maximum performance available. But it might be a waste of resources and you might not really need to handle 30,000 (or more) requests per second. In that case, you can play around with those configurations as you need.

There are several combinations of configurations you can use, check the docs: https://github.com/tiangolo/uvicorn-gunicorn-fastapi-docker

**Comment by SansDK** [comment]
Hey @tiangolo, thanks for your reply.

I tried to set the ```WEB_CONCURRENCY``` variable (to 2), but still I'm getting the following log:
```
[2019-05-24 14:15:29 +0000] [488] [INFO] Booting worker with pid: 488
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
[2019-05-24 14:15:37 +0000] [488] [INFO] Started server process [488]
[2019-05-24 14:15:37 +0000] [488] [INFO] Waiting for application startup.
[2019-05-24 14:15:38 +0000] [493] [INFO] Booting worker with pid: 493
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
[2019-05-24 14:15:46 +0000] [493] [INFO] Started server process [493]
[2019-05-24 14:15:46 +0000] [493] [INFO] Waiting for application startup.
[2019-05-24 14:15:47 +0000] [498] [INFO] Booting worker with pid: 498
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
[2019-05-24 14:15:54 +0000] [498] [INFO] Started server process [498]
[2019-05-24 14:15:54 +0000] [498] [INFO] Waiting for application startup.
[2019-05-24 14:15:56 +0000] [503] [INFO] Booting worker with pid: 503
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
[2019-05-24 14:16:03 +0000] [503] [INFO] Started server process [503]
[2019-05-24 14:16:03 +0000] [503] [INFO] Waiting for application startup.
[2019-05-24 14:16:05 +0000] [508] [INFO] Booting worker with pid: 508
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
[2019-05-24 14:16:13 +0000] [508] [INFO] Started server process [508]
[2019-05-24 14:16:13 +0000] [508] [INFO] Waiting for application startup.
[2019-05-24 14:16:14 +0000] [513] [INFO] Booting worker with pid: 513
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
[2019-05-24 14:16:22 +0000] [513] [INFO] Started server process [513]
[2019-05-24 14:16:22 +0000] [513] [INFO] Waiting for application startup.
[2019-05-24 14:16:23 +0000] [518] [INFO] Booting worker with pid: 518
WARNING:root:email-validator not installed, email fields will be treated as str.
```

Would you have any idea why this could be happening with the variable set to a fixed, small value?
Could an error in my Python code cause this?

I am currently executing some code in this method:
```
@app.on_event("startup")
def startup_event():
    ...
```

Could it be that the application keeps trying to start if something in here fails?

Thanks.

**Comment by tiangolo** [comment]
That's quite strange. I guess it might be related to the startup event, if there's an exception there, I think it won't necessarily be properly handled by Uvicorn (specifically there in the startup event), I think there's an issue about it somewhere, but I'm not sure.

Could you put a `try` there and log if there's any exception? That might help finding out the root case of the problem.

**Comment by SansDK** [comment]
That's what I thought too.

Unfortunately I won't be able to try this out anymore today, but I'll come back on this as the first thing on monday morning. 

Thanks so far!

**Comment by tiangolo** [comment]
Cool!

**Comment by SansDK** [comment]
Hi there, I'm back.
I added a try/except, but there is no exception while running.
Now, I commented out everything in (and including) the:
```
@app.on_event("startup")
def startup_event():
```

And now it runs fine..

Things that I _was_ doing in the startup method were:

```
global db_ip, db_port, x, y, z
x = load_x()   # Reads a pickled file and returns it
y = load_y()   # Reads a pickled file and returns it
db_ip = os.getenv('DB_IP', default="...")
db_port = os.getenv('DB_PORT', default="...")
z = init_z()     # Makes a connection to the DB and executes a query fairly quickly
```

Is any of this not possible, or not smart to do in the startup method?

Thanks.

**Comment by euri10** [comment]
you didn't print the whole gunicorn log, in particular the default gunicorn conf spits out the number of workers from [here](https://github.com/tiangolo/uvicorn-gunicorn-docker/blob/dea28f070ea5bfcb9dda6b6f66d8708562a1f861/python3.7-alpine3.8/gunicorn_conf.py#L32-L42)


If after checking it's effectively well set at 2 because you configured with WEB_CONCURRENCY, what I'd do is use a custom gunicorn conf, like explained [here](https://github.com/tiangolo/uvicorn-gunicorn-docker#custom-gunicorn-configuration-file)  

but imho you're hit by https://github.com/tiangolo/uvicorn-gunicorn-docker/blob/dea28f070ea5bfcb9dda6b6f66d8708562a1f861/python3.7-alpine3.8/gunicorn_conf.py#L23

in my custom gunicorn config I use something like
```
max_workers = os.getenv("MAX_WORKERS", "2")
...
    web_concurrency = min(int(max_workers), int(default_web_concurrency))

```

I think the 1st thing is to make sure you effectively launch gunicorn with 2 workers, from the log it seems not, then after look further down

**Comment by SansDK** [comment]
Are you sure that the problem is related to that?

If I comment out my startup event code, this is the (full) log that I'm getting:
```
Checking for script in /app/prestart.sh
Running script /app/prestart.sh
Running inside /app/prestart.sh, you could add migrations to this file, e.g.:
 #! /usr/bin/env bash
 # Let the DB start
sleep 10;
# Run migrations
alembic upgrade head
 [2019-05-27 06:50:44 +0000] [1] [INFO] Starting gunicorn 19.9.0
[2019-05-27 06:50:44 +0000] [1] [INFO] Listening at: http://0.0.0.0:8081 (1)
[2019-05-27 06:50:44 +0000] [1] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2019-05-27 06:50:44 +0000] [8] [INFO] Booting worker with pid: 8
[2019-05-27 06:50:44 +0000] [9] [INFO] Booting worker with pid: 9
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
WARNING:root:email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
[2019-05-27 06:50:53 +0000] [8] [INFO] Started server process [8]
[2019-05-27 06:50:53 +0000] [8] [INFO] Waiting for application startup.
[2019-05-27 06:50:53 +0000] [9] [INFO] Started server process [9]
[2019-05-27 06:50:53 +0000] [9] [INFO] Waiting for application startup.
```

And that's all. No looping like in the previous logs.
When I shut the instance down, I get the gunicorn debug log like you said. It looks like this:
```
{"loglevel": "info", "workers": 2, "bind": "0.0.0.0:8081", "workers_per_core": 1.0, "host": "0.0.0.0", "port": "8081"}
```

This confirms that there are 2 workers, and just 1 worker per core.

To me it seems more like there is something weird going on in the startup event. I couldn't find anything about the startup event method in this repo, so I'm not sure how it works exactly. Is there some timeout value on the startup method, maybe?

Thanks for the help.

**Comment by euri10** [comment]
so if I got it correctly, WITHOUT a custom `@app.on_event("startup")` like below 
```
global db_ip, db_port, x, y, z
x = load_x()   # Reads a pickled file and returns it
y = load_y()   # Reads a pickled file and returns it
db_ip = os.getenv('DB_IP', default="...")
db_port = os.getenv('DB_PORT', default="...")
z = init_z()     # Makes a connection to the DB and executes a query fairly quickly
```
it works fine, 2 workers like you set

Now should you add that code above it spins more workers ?

a minimal example would help understand what happens

**Comment by SansDK** [comment]
Something like that seems to be happening, yes. Maybe because of some error happening in the startup, but it doesn't throw any exception.

I can't share the full code, but I'll provide the baseline.
main.py:
```
import pandas as pd
import pickle
import xgboost as xgb
# import uvicorn
import psycopg2 as pg
import os
import importlib
import sys
import json
from sklearn.preprocessing import QuantileTransformer
from typing import List
from datetime import datetime
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI(title="Title", description="Description ", version="v0.1")

@app.post("/x", tags=["tag1"], summary="Summary")
def x_func(req: Request):
# couple of more endpoints and methods

def load_x():
    try:
        return pickle.load(open("filex.dat", "rb"))
    except:
        print("Can't load x file", file=sys.stderr)
        return None

def load_y():
    try:
        return pickle.load(open("filey.dat", "rb"))
    except:
        print("Can't load y file", file=sys.stderr)
        return None

def init_z():
    connection = pg.connect("..........")
    return pd.read_sql("SELECT ........")

db_ip = ""
db_port = ""
x = None
y = None
z = None

@app.on_event("startup")
def startup_event():
    global db_ip, db_port, x, y, z
    x = load_x()   # Reads a pickled file and returns it
    y = load_y()   # Reads a pickled file and returns it
    db_ip = os.getenv('DB_IP', default="...")
    db_port = os.getenv('DB_PORT', default="...")
    z = init_z()     # Makes a connection to the DB and executes a query fairly quickly
```

Dockerfile:
```
FROM tiangolo/uvicorn-gunicorn-fastapi:python3.6

RUN pip install pandas numpy requests fastapi pydantic datetime xgboost psycopg2 scikit-learn==0.20.3

ENV DB_IP=...
ENV DB_PORT=...

COPY ./app /app
```

Kubernetes yaml file:
```
      - name: test-name
        imagePullPolicy: Always
        image: ...image
        env:
        - name: PORT
          value: '8081'
        - name: WEB_CONCURRENCY
          value: '2'
        ports:
        - containerPort: 8081
        resources:
          limits:
            cpu: 200m
            memory: 300Mi
          requests:
            cpu: 50m
            memory: 150Mi
```

Is that enough of an example or do you need other more specific info?
Thanks!

**Comment by SansDK** [comment]
Hey,

I did some further research and I found out what is causing the issue.
It seems to be some kind of timeout issue. I was messing around commenting and uncommenting lines of code, and everything worked without executing the DB query. Then I uncommented the DB query and added a LIMIT, which also worked!

So with the when executing the following query, everything works fine (and gunicorn just boots 2 workers):
```
pd.read_sql("SELECT ... LIMIT 1000", connection)
```

And when I change the query to no limit, gunicorn keeps booting workers (which has something to do with some kind of startup timeout I guess):
```
pd.read_sql("SELECT ...", connection)
```

Without the LIMIT it's fetching around 185k rows with 5 columns, which is quite a bit, but not huge.

@euri10 @tiangolo Do any of you know a fix for this problem? Can I define a timeout value in a gunicorn conf file maybe?

Thank you.

**Comment by tiangolo** [comment]
You can add a custom `timeout` variable to a custom Gunicorn config. But I think I should add an env var for this, it seems to be a frequent need.

**Comment by SansDK** [comment]
Thanks a lot @tiangolo, that fixed my problem!
I added a ```gunicorn_conf.py``` file and added:
```
timeout = 100
graceful_timeout = 100
```

An env var for this would be a nice to have, but since the gunicorn conf is easy to use as well, I wouldn't give it a high priority.

Thanks again for the help, and thanks for the great FastAPI lib!

**Comment by SansDK** [comment]
Sorry, I spoke too quickly.
I still had the ```LIMIT 1000``` in place, so of course it worked (I haven't had my coffee yet).
Now that I removed the limit, it keeps booting new workers like earlier, sigh.

Guess I'll try increasing the timeout value, but I highly doubt that it can even take longer than 100 seconds already.

**Comment by SansDK** [comment]
I set the timeout value to 200 seconds, but still no succes.
The weird thing is that this starts happening in way less than 200 seconds, which indicates that it

- ignores the timeout (unlikely)
- or, the env vars don't get passed on through Kubernetes correctly somehow

I'll do some checks to see if the env vars get passed on correctly.

**Comment by SansDK** [comment]
For now I changed ```WEB_CONCURRENCY``` to 1, which is not ideal but at least it seems to prevent the continuous worker-booting.

**Comment by tiangolo** [comment]
It's strange that it works with `WEB_CONCURRENCY` 1 but doesn't with more workers, might be some other configuration in Gunicorn or something else in the environment...?

Anyway, I'm glad you solved your problem.

An alternative that might be equally feasible (or maybe even better) is to replicate the containers, taking advantage of Docker and Kubernetes. That also means that you can scale more easily in memory and processors horizontally, by not being constrained to a single container.

**Comment by davepaiva** [comment]
I had the same issue, running FASTApi server on ECS. Turns out the timeout was because by server could not access the RDS on another VPC due to Sercurity Group Rules.

**Comment by weiyiyin0321** [comment]
I'm having the same issue as well. My api works with `WEB_CONCURRENCY="1"` but when I increase the number of workers, it goes into a loop of booting up new workers over and over again. I've increased the `timeout` configuration in `gunicorn_conf.py` but doesn't seem to help. For now, I will scale by replicating containers, but would be helpful if we can get a solution to this.

**Comment by fdroessler** [comment]
I ran into the same issue. I am suspecting in my case (8 cpus) with webconcurrency at None and workers per core 1 by default it would try to start 8 workers. However, the data I am loading at on_startup is probably to large for 8 workers so maybe they die and gunicorn tries to restart those who crash. Does that make sense?

**Comment by tiangolo** [comment]
Yeah, I think that could make sense @fdroessler .

**Comment by ysde** [comment]
Hi 
I have the same issues when deploy FastAPI on GCP's k8s.

Can't figure what the problem is.

Thanks in advance

**Comment by ysde** [comment]
> I have the same issues when deploy FastAPI on GCP's k8s.

It turns out is the pod's memory limit is too small, after increase the memory limit the `booting worker` issue disappear it.

**Comment by MchlUh** [comment]
Just posting my experience on something similar : 
I was using some app.on_event("startup") logic, on a DigitalOcean VM with shared vCPUs.
I had workers keeping rebooting, and fixed this by switching my deployment to dedicated vCPUs

**Comment by RiON69** [comment]
At the time of the start, the gunicorn spends a lot of CPU resources to raise the workers. Rather, the CPU limit kills the workers by causing an infinite loop.
For example, in my case with 8 workers and settings:
```
resources:
  limits:
    cpu: 1
    memory: 1Gi
  requests:
    cpu: 0.1
    memory: 128Mi
```
causes the workers to restart. It can be seen that the load on the processes reaches 1000m.

Therefore, the solution can be either to reduce the number of workers, or to increase the processor limit.
